02_bubble/saurabhvyas9/main.go

```diff
@@ -0,0 +1,44 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(s []int) []int {
+	r := make([]int, len(s))
+	copy(r, s)
+	countChecked := 1
+	for countChecked != 0 {
+		countChecked = 0
+		for i := 0; i < (len(r) - 1); i++ {
+			if r[i] > r[i+1] {
+				r[i], r[i+1] = r[i+1], r[i]
+				countChecked++
+			}
+		}
+	}
+	return r
+}
+
+func insertion(s []int) []int {
+	r := make([]int, len(s))
+	copy(r, s)
+	for i := 1; i < len(r); i++ {
```

Comment:
It took me a while to get my head around this - but it does seem to do insertion sort:
build up a sorted array from the left to the right.
However, it does more comparisons than necessary (the second iteration should go from i down and break when the insertion point is found). Also, the `temp` slice is not needed and does not add to readability IMO.

So a first approach to simplify this would be as:
```go
	for i := 1; i < len(r); i++ {
		for j := 0; j < i; j++ {
			if r[j] > r[i] {
				r[j], r[i] = r[i], r[j]
			}
		}
	}
```

Secondly, start the inner loop from `i` and decrement and also add an early bail once the current element to be inserted is bigger than the remaining elements to the left (what ever is left of the element to be inserted _is already sorted_)
```go
	for i := 1; i < len(s); i++ {
		for j := i; j > 0 && s[j-1] > s[j]; {
			s[j-1], s[j] = s[j], s[j-1]
			j--
		}
	}
	return s
``` 
this is short and sweet, but it does it does twice as many writes as necessary and it is also not straight obvious how the `s[i]` is the current element to be inserted.
The following listing might make this clearer:

```go
	for i := 1; i < len(s); i++ {
		var j int
		for j = i; j > 0; j-- {
			if s[j-1] <= s[i] {
				break // We've found the insertion point
			}
		}
		// insert s[i] at s[j] moving the rest up
		tmp := s[i]
		copy(s[j+1:i+1], s[j:i])
		s[j] = tmp
	}
```

It's up to you if you wan to update it.

Reply:
Thanks Julia. Really Appreciate your feedback. 
I am using `temp` slice because it helped me visualise the slice am working on.
But I got your point about the logic and that the usage of loop was more than required. With the recent commit, I have reduced it much more. (Now, loop usage is even less than suggested logic)
Thank you helping me improve. :)